Lib/typing.py

```diff
@@ -930,6 +945,14 @@ def __subclasscheck__(self, cls):
                 return True
 
 
+class _LiteralGenericAlias(_GenericAlias, _root=True):
+
+    def __eq__(self, other):
+        if not isinstance(other, _LiteralGenericAlias):
+            return NotImplemented
+
+        return len(self.__args__) == len(other.__args__) and set(self.__args__) == set(other.__args__)
```

Comment:
One problem with this:
```python
>>> Literal[1, 2, 3] == Literal[1, 2, 3, 3]
False
```
According to [the pep](https://www.python.org/dev/peps/pep-0586/#shortening-unions-of-literals):
> Literal[v1, v2, v3] is equivalent to Union[Literal[v1], Literal[v2], Literal[v3]]

Union de-duplicates args, so IMO, Literal should do this too. There are two approaches possible for this:
1. Change the return in `__eq__` to just:
```suggestion
        set(self.__args__) == set(other.__args__)
```
This preserves the original `__args__`.

2. Or just return `set(params)` instead of `tuple(params)` in `_flatten_literal_params`.
This doesn't preserve the original `__args__`, but is how Union handles it.

Reply:
Thanks for the suggestions.

First option can't be used because `Literal[0] == Literal[False]` will be `True` and peps says:
>  Literal[0] and Literal[False] is not equivalent despite that 0 == False evaluates to 'true' at runtime: 0 has type int and False has type bool.

Regarding the second option, when you trying to create `Literal` with mutable type, for instance `Literal[{1: 2}]`, it should not fail at runtime, so if we will try to create `Literal[{1: 2}]` it will fail because the parameter is not hashable.

I have updated the pull request. Basically, it will construct a set of pairs (parameter and parameter type) for literals and then compare those sets. So code like this `Literal[False] == Literal[0]` will be `False`.

@Fidget-Spinner what is your opinion regarding such solution?

Reply:
Oops, you're right about both points, I forgot about the special `{0} == {False}`, and that there can be non-hashable args for Literals.

I think this is a good solution, thanks.